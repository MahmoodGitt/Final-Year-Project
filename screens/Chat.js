/* eslint-disable react/prop-types */
import React, { useState, useCallback, useLayoutEffect } from 'react';

import { GiftedChat } from 'react-native-gifted-chat';
import {
	orderBy,
	collection,
	query,
	getFirestore,
	onSnapshot,
	addDoc,
} from 'firebase/firestore';
import auth from '../firebase/config';

const Chat = ({ route }) => {
	const [messages, setMessages] = useState([]);
	const chatroomId = route.params.chatroomId;

	useLayoutEffect(() => {
		const loadMessages = () => {
			const db = getFirestore();
			const q = query(
				collection(db, 'chatrooms/' + chatroomId + '/messages'),
				orderBy('createdAt', 'desc')
			);
			const unsubscribe = onSnapshot(q, (snapshot) =>
				setMessages(
					snapshot.docs.map((doc) => ({
						_id: doc.data()._id,
						createdAt: doc.data().createdAt.toDate(),
						text: doc.data().text,
						user: doc.data().user,
					}))
				)
			);
			return () => {
				unsubscribe();
			};
		};
		loadMessages();
	}, []);

	const onSend = useCallback((msg = []) => {
		const { _id, createdAt, text, user } = msg[0];
		const database = getFirestore();
		addDoc(collection(database, 'chatrooms/' + chatroomId + '/messages'), {
			_id,
			createdAt,
			text,
			user,
		});

		setMessages((prevMessages) => GiftedChat.append(prevMessages, msg));
	}, []);

	return (
		<GiftedChat
			messages={messages}
			onSend={(messages) => onSend(messages)}
			user={{
				_id: auth?.currentUser?.uid,
				avatar: 'https://placeimg.com/140/140/any',
			}}
			showAvatarForEveryMessage={true}
		/>
	);
};

export default Chat;
