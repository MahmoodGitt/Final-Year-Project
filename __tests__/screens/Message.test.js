/* eslint-disable no-undef */
import React from 'react';
import renderer from 'react-test-renderer';

import Chat from '../../screens/Chat';

describe('Creat Chat Component', () => {
	it('Testing if Message Component loads up', () => {
		const tree = renderer.create(<Chat />).toJSON();
		expect(tree).toMatchSnapshot();
	});
});
