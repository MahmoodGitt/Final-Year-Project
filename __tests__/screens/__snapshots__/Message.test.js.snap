// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creat Chat Component Testing if Message Component loads up 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;
